src/webgpu/util/texture/texelData.ts

```diff
@@ -66,21 +75,28 @@ const sint = (bitLength: number) => (n: number) => ({
   type: TexelWriteType.Sint,
 });
 
-const unorm2 = { write: unorm(2), bitLength: 2 };
-const unorm8 = { write: unorm(8), bitLength: 8 };
-const unorm10 = { write: unorm(10), bitLength: 10 };
+const decodeUnorm = (bitLength: number) => (n: number) =>
+  normalizedIntegerAsFloat(n, bitLength, false);
+const decodeSnorm = (bitLength: number) => (n: number) =>
+  normalizedIntegerAsFloat(n, bitLength, true);
+const identity = (n: number) => n;
+
+const unorm2 = { decode: decodeUnorm(2), write: unorm(2), bitLength: 2 };
+const unorm8 = { decode: decodeUnorm(8), write: unorm(8), bitLength: 8 };
+const unorm10 = { decode: decodeUnorm(10), write: unorm(10), bitLength: 10 };
 
-const snorm8 = { write: snorm(8), bitLength: 8 };
+const snorm8 = { decode: decodeSnorm(8), write: snorm(8), bitLength: 8 };
 
-const uint8 = { write: uint(8), bitLength: 8 };
-const uint16 = { write: uint(16), bitLength: 16 };
-const uint32 = { write: uint(32), bitLength: 32 };
+const uint8 = { decode: identity, write: uint(8), bitLength: 8 };
+const uint16 = { decode: identity, write: uint(16), bitLength: 16 };
+const uint32 = { decode: identity, write: uint(32), bitLength: 32 };
 
-const sint8 = { write: sint(8), bitLength: 8 };
-const sint16 = { write: sint(16), bitLength: 16 };
-const sint32 = { write: sint(32), bitLength: 32 };
+const sint8 = { decode: identity, write: sint(8), bitLength: 8 };
+const sint16 = { decode: identity, write: sint(16), bitLength: 16 };
+const sint32 = { decode: identity, write: sint(32), bitLength: 32 };
 
 const float10 = {
+  decode: identity,
```

Comment:
Ideally, this wouldn't actually be the identity function but would cut the precision of the input value down so that it could be represented by a float10

I'm not sure if that will actually help the precision issues though or simply need a better way to compute tolerance.